# DATA CLEANING CHALLENGE #MARCH

## Introduction 

DataCleaning challenge provides an opportunity for every Data Analyst at all levels of expertise (beginner, intermediate,  or even expert) to build a portfolio-worthy project that can be shared with recruiters.

The challenge also provides an avenue for Data Analysts to meet with fellow learners and build a great network.   

#### What to look out for when cleaning your data

 - Incorrect data types
 - Null entries
 - Missing values
 - Duplicate entries
 - Errors in spellings and values
 - Wrong calculations across rows and columns
 - Irrelevant data 
 - Outliers 

First, importing the necessary libraries required to perform the above operations

 - ## Importing Libraries

In [1]:
#Importing Libraries 
import pandas as pd
import numpy as np

In [2]:
# read the data set into dataframe using the pandas library
pd.set_option('display.max_columns', None)
data = pd.read_csv('fifa21 raw data v2.csv', skip_blank_lines=True, low_memory=False,)


Lets view our dataframe and observe the data

 - ## Observing The Dataset

In [3]:
# previewing the first 10 rows of the dataset
data.head(10)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   
5  188545     R. Lewandowski            Robert Lewandowski   
6  209331           M. Salah                 Mohamed Salah   
7  212831            Alisson         Alisson Ramses Becker   
8  231747          K. Mbappé                 Kylian Mbappé   
9  192448      M. ter Stegen         Marc-André ter Stegen   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   
5  https://cdn.sofifa.com/players/188/545/21_60.png   
6  https://cdn.sofifa.com/players/209/331/21_60.png   
7  https://cdn.sofifa.com/players/212/831/21_60.png   
8  https://cdn.sofifa.com/players/231/747/21_60.png   
9  https://cdn.sofifa.com/players/192/448/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   
5  http://sofifa.com/player/188545/robert-lewando...      Poland   31    91   
6  http://sofifa.com/player/209331/mohamed-salah/...       Egypt   28    90   
7  http://sofifa.com/player/212831/alisson-ramses...      Brazil   27    90   
8  http://sofifa.com/player/231747/kylian-mbappe/...      France   21    90   
9  http://sofifa.com/player/192448/marc-andre-ter...     Germany   28    90   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   
5   91    \n\n\n\nFC Bayern München  2014 ~ 2023          ST  184cm   80kg   
6   90            \n\n\n\nLiverpool  2017 ~ 2023          RW  175cm   71kg   
7   91            \n\n\n\nLiverpool  2018 ~ 2024          GK  191cm   91kg   
8   95  \n\n\n\nParis Saint-Germain  2018 ~ 2022  ST, LW, RW  178cm   73kg   
9   93         \n\n\n\nFC Barcelona  2014 ~ 2022          GK  187cm   85kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   
5          Right   91            ST   Jul 1, 2014           NaN    €111M   
6           Left   90            RW   Jul 1, 2017           NaN  €120.5M   
7          Right   90            GK  Jul 19, 2018           NaN    €102M   
8          Right   91            ST   Jul 1, 2018           NaN  €185.5M   
9          Right   90            GK   Jul 1, 2014           NaN    €110M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70  

In [4]:
# to see the numbers of rows and columns we have in our dataset
data.shape

(18979, 77)

we have 18979 rows and 77 columns in the dataset

lets see the names of the columns we have

In [5]:
# the preview the name of the columns we have in the dataset
data.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

##### Obeservation
It was observed that some columns do not follow the standard naming convention, some of the column names where separate with space instead of " _ " 

In [6]:
# renaming the columns and replacing white spaces with an underscore
data.columns = [col.replace(" ", "_") for col in data.columns]

lets see if our change is applied

In [7]:
# previewing the column once again to see if there is any further adjustment needed
data.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred_Foot', 'BOV', 'Best_Position', 'Joined', 'Loan_Date_End',
       'Value', 'Wage', 'Release_Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading_Accuracy', 'Short_Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK_Accuracy', 'Long_Passing', 'Ball_Control', 'Movement',
       'Acceleration', 'Sprint_Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot_Power', 'Jumping', 'Stamina', 'Strength', 'Long_Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing_Tackle',
       'Sliding_Tackle', 'Goalkeeping', 'GK_Diving', 'GK_Handling',
       'GK_Kicking', 'GK_Positioning', 'GK_Reflexes', 'Total_Stats',
       'Base_Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

Lets see some other features of the data

The information method contains the number of columns, column labels, column data types, memory usage, range index, and the number of cells in each column (non-null values).

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred_Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best_Position     18979 non-null  object
 17  Joined      

Some of the columns are abbreviated the meaning can be found [here](dataset_dictionary.txt). 

Some columns name are not well represented, lets fix that

#### Observation
* The OVA column is not well represented
* The Height and Weight column are of the object type
* Some columns like Loan_Date_End, Hits contains NaN (Null) values

 - ## Cleaning The Data

In [9]:
#the OVA columns is not well spelt
# renaming that particular column is ideal
data.rename(columns = {'↓OVA': 'OVA'}, inplace = True)

In [10]:
# Previewing the columns again to check if the change is applied
data.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       'OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred_Foot', 'BOV', 'Best_Position', 'Joined', 'Loan_Date_End',
       'Value', 'Wage', 'Release_Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading_Accuracy', 'Short_Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK_Accuracy', 'Long_Passing', 'Ball_Control', 'Movement',
       'Acceleration', 'Sprint_Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot_Power', 'Jumping', 'Stamina', 'Strength', 'Long_Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing_Tackle',
       'Sliding_Tackle', 'Goalkeeping', 'GK_Diving', 'GK_Handling',
       'GK_Kicking', 'GK_Positioning', 'GK_Reflexes', 'Total_Stats',
       'Base_Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', '

Now, the column has been successfully renamed

Going back to the data.info() to observe the dataset

We also observed some of the columns contains NAN values or empty values

In [11]:
# Preview the whole dataset for further observation
data.head(5)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred_Foot  BOV Best_Position        Joined Loan_Date_End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vi

#### Another Observation
It was observed that the Club column has some unwanted strings in it

In [12]:
#Lets preview the Club column alone
data.Club

0               \n\n\n\nFC Barcelona
1                   \n\n\n\nJuventus
2            \n\n\n\nAtlético Madrid
3            \n\n\n\nManchester City
4        \n\n\n\nParis Saint-Germain
                    ...             
18974             \n\n\n\nWuhan Zall
18975        \n\n\n\nOldham Athletic
18976             \n\n\n\nDerry City
18977       \n\n\n\nDalian YiFang FC
18978       \n\n\n\nDalian YiFang FC
Name: Club, Length: 18979, dtype: object

Let remove the unwanted string in the column

In [13]:
# Previewing how the data set will look like if we remove the unwanted char
print(data['Club'].str[4:])

0               FC Barcelona
1                   Juventus
2            Atlético Madrid
3            Manchester City
4        Paris Saint-Germain
                ...         
18974             Wuhan Zall
18975        Oldham Athletic
18976             Derry City
18977       Dalian YiFang FC
18978       Dalian YiFang FC
Name: Club, Length: 18979, dtype: object


In [14]:
# Using the unique mechod to get the unique values present in the dataset
# The Unique value is use to observe if we have abnormalities in that particular column
data['Club'].unique()

array(['\n\n\n\nFC Barcelona', '\n\n\n\nJuventus',
       '\n\n\n\nAtlético Madrid', '\n\n\n\nManchester City',
       '\n\n\n\nParis Saint-Germain', '\n\n\n\nFC Bayern München',
       '\n\n\n\nLiverpool', '\n\n\n\nReal Madrid', '\n\n\n\nChelsea',
       '\n\n\n\nTottenham Hotspur', '\n\n\n\nInter', '\n\n\n\nNapoli',
       '\n\n\n\nBorussia Dortmund', '\n\n\n\nManchester United',
       '\n\n\n\nArsenal', '\n\n\n\nLazio', '\n\n\n\nLeicester City',
       '\n\n\n\nBorussia Mönchengladbach', '\n\n\n\nReal Sociedad',
       '\n\n\n\nAtalanta', '\n\n\n\nOlympique Lyonnais', '\n\n\n\nMilan',
       '\n\n\n\nVillarreal CF', '\n\n\n\nRB Leipzig', '\n\n\n\nCagliari',
       '\n\n\n\nAjax', '\n\n\n\nSL Benfica', '\n\n\n\nAS Monaco',
       '\n\n\n\nWolverhampton Wanderers', '\n\n\n\nEverton',
       '\n\n\n\nFiorentina', '\n\n\n\nFC Porto', '\n\n\n\nRC Celta',
       '\n\n\n\nTorino', '\n\n\n\nSevilla FC', '\n\n\n\nGrêmio',
       '\n\n\n\nReal Betis', '\n\n\n\nRoma', '\n\n\n\nNewcastle Unite

#### Obersvation
from the Observation, the string char "\n\n\n\n" is common to most of the string characters in that column
we can replace the unwanted string char with an empty string

In [15]:
# replacing the unwanted string character with an empty string
data['Club'] = data['Club'].str.replace('\n\n\n\n', '')

In [16]:
# lets inspect the dataset to see if the change is applied
data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   91   

   POT                 Club     Contract   Positions Height Weight  \
0   93         FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             Juventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      Atlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      Manchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  Paris Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred_Foot  BOV Best_Position        Joined Loan_Date_End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44     

#### Further cleaning
Also, for our Height and Weight, We can rename the columns to carry the SI unit, then removing the unit from the entire column, and then convert the column to an int type

In [17]:
# renaming the Height column to by adding the SI of the height to it

data.rename(columns = {'Height': 'Height_cm'}, inplace = True)

In [18]:
# renaming the Weight column to by adding the SI of weight to it

data.rename(columns = {'Weight': 'Weight_kg'}, inplace = True)

In [19]:
# Inspecting the dataset to see if the change is applied
data.head(4)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   

                                           playerUrl Nationality  Age  OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   91   

   POT             Club     Contract   Positions Height_cm Weight_kg  \
0   93     FC Barcelona  2004 ~ 2021  RW, ST, CF     170cm      72kg   
1   92         Juventus  2018 ~ 2022      ST, LW     187cm      83kg   
2   93  Atlético Madrid  2014 ~ 2023          GK     188cm      87kg   
3   91  Manchester City  2015 ~ 2023     CAM, CM     181cm      70kg   

  Preferred_Foot  BOV Best_Position        Joined Loan_Date_End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   

    Wage Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   
2        140          34             19           11      65         11   
3        408          76             66           88      94         84   

   Composure  Defending  Marking  Standing_Tackle  Sliding_Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   
2         68         57       27               12              18   
3         91        186       68               65              53   

   Goalkeeping  GK_Di

Since the Height column has been renamed, we can now remove the SI from the entire column.
before then, lets see if there are some abnormal values in the column

### Cleaning the Height column

In [20]:
# using the unique() method to see the unique values contained in the data set
data['Height_cm'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

From our observation, not all the values in the height column are represented in centimeters(cm),
We will remove the cm from all the values in the column 
also, we will clean the abnormal values which was represented in ft-inches and then convert it to cm

In [21]:
# replace the values that are represented in ft, inches and converting it to centimetres

In [22]:
# observing the unique values in the column before conversion
data['Height_cm'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

In [23]:
data.Height_cm.sample(10)

15084    185cm
3325     182cm
10638    189cm
9548     176cm
17923    183cm
1315     192cm
6455     178cm
9550     172cm
9942     179cm
10974    185cm
Name: Height_cm, dtype: object

In [24]:
#to define a function that will do the conversion, i will be using the unique last value as a reference
data['Height_cm'].str[-1].unique()

array(['m', '"'], dtype=object)

In [25]:
# define a function 
#using the unique last value as a reference of condition
# replacing the unwant string value with an empty string
# converting the concerned value from their respective units to centimeter
# Adding both units to complete the conversion
# return the answer
# if else, resume the other centimeter value without the cm
def convert_height(x):
    if x[-1] == '"':
        x = x.replace("\"","")
        inch = int(x[2:]) * 2.54
        foot = int(x[0]) * 30.48
        return (foot+inch)
    elif x[-1] == "m":
        return int(x[:-2])


In [26]:
# Applying the function to the Height column
data['Height_cm'] = data['Height_cm'].apply(convert_height)

In [27]:
#checking if the change is applied
data['Height_cm'].unique()

array([170.  , 187.  , 188.  , 181.  , 175.  , 184.  , 191.  , 178.  ,
       193.  , 185.  , 199.  , 173.  , 168.  , 176.  , 177.  , 183.  ,
       180.  , 189.  , 179.  , 195.  , 172.  , 182.  , 186.  , 192.  ,
       165.  , 194.  , 167.  , 196.  , 163.  , 190.  , 174.  , 169.  ,
       171.  , 197.  , 200.  , 166.  , 187.96, 164.  , 198.  , 190.5 ,
       195.58, 180.34, 193.04, 185.42, 182.88, 177.8 , 175.26, 167.64,
       170.18, 162.56, 201.  , 158.  , 162.  , 161.  , 160.  , 203.  ,
       157.  , 156.  , 202.  , 159.  , 206.  , 155.  ])

In [28]:
data['Height_cm'].isnull().sum()

0

### Cleaning the weight column
Also for the Weight column, we will convert the lbs to kg

Also for the weight column

In [29]:
#previewing the unique values of the weight column the SI units are both kg and lbs
data['Weight_kg'].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

In [30]:
#to define a function that will do the conversion, i will be using the unique last value as a reference
data['Weight_kg'].str[-1].unique()

array(['g', 's'], dtype=object)

In [31]:
# define a function 
#using the unique last value as a reference of condition
# replacing the unwant string value with an empty string
# converting the concerned value from it unit to Kilogram
# return the rounded off answer
# if else, preserve the other kilogram value without the kg

def convert_weight(x):
    if x[-1] == 's':
        x = x.replace("lbs","")
        kg = int(x[:3]) / 2.205
        return round(kg)
    elif x[-1] == "g":
        return int(x[:-2])

In [32]:
# Applying the function to the weight column 
data['Weight_kg'] = data['Weight_kg'].apply(convert_weight)

In [33]:
# checking to see if the change is applied
data['Weight_kg'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52])

In [34]:
#Previewing the data set
data.head(5)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   91   

   POT                 Club     Contract   Positions  Height_cm  Weight_kg  \
0   93         FC Barcelona  2004 ~ 2021  RW, ST, CF      170.0         72   
1   92             Juventus  2018 ~ 2022      ST, LW      187.0         83   
2   93      Atlético Madrid  2014 ~ 2023          GK      188.0         87   
3   91      Manchester City  2015 ~ 2023     CAM, CM      181.0         70   
4   91  Paris Saint-Germain  2017 ~ 2022     LW, CAM      175.0         68   

  Preferred_Foot  BOV Best_Position        Joined Loan_Date_End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vi

## Moving on with cleaning other columns

Also according to the info() of the dataset, some other columns needs cleaning

### Cleaning The Contract Column

In [35]:
# Checking the unique values in the contract column
data['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

Its is observed that some players have year range contracts which includes the date and year of expiration,
Some players are on loan with year of expiration
Some players are on Free contracts

So Idealy,it is ideal to convert the column into a categorical column, i will be using the last value on each column as a reference for my python function

In [36]:
#Some row with the year range contains the range of 0 - 9 as their last string value
str(list(range(10)))

'[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]'

In [37]:
# see the unique last value of the contract column
data['Contract'].str[-1].unique()

array(['1', '2', '3', '4', '5', '6', 'n', 'e', '8', '0', '7'],
      dtype=object)

In [38]:
#Using a function to set values for each rows containing the unique last values used as reference

def convert_Cont(x):
    if x[-1] in str(list(range(10))):
        x = 'Contract'
        return x
    elif x[-1] == 'n':
        x = 'Loan'
        return x
    elif x[-1] == 'e':
        x = 'Free'
        return x
        

In [39]:
#Applying the function
data['Contract'] = data['Contract'].apply(convert_Cont)

In [40]:
#converting the column into a categorical type 
data['Contract'] = data['Contract'].astype('category')

In [41]:
#renaming the column from contract to contract status
data.rename(columns = {'Contract': 'Contract_Status'}, inplace = True)

In [42]:
data[data['Contract_Status'] == 'Free']

ID            Name                          LongName  \
289    230347  Welington Dano    Welington Kauê Dano Nascimento   
292    230225  Juiano Mestres  Juan Everton Mestres de Mesquita   
369    245299     J. Frendado            Jaime Nicolás Frendado   
374    245294    J. Serendero          Jorge Ezequiel Serendero   
375    245308        M. Nérez                Mauro Evidio Nérez   
...       ...             ...                               ...   
17220  247059   A. Al Hidhani                    Ali Al Hidhani   
17343  251734       A. Sanghu                    Anudaan Sanghu   
17659  251738        M. Chada                   Madanapal Chada   
17661  251741       E. Suresh                Eelamynthan Suresh   
18205  251737       J. Mangat                  Jagatjiva Mangat   

                                               photoUrl  \
289    https://cdn.sofifa.com/players/230/347/21_60.png   
292    https://cdn.sofifa.com/players/230/225/21_60.png   
369    https://cdn.sofifa.com/players/245/299/21_60.png   
374    https://cdn.sofifa.com/players/245/294/21_60.png   
375    https://cdn.sofifa.com/players/245/308/21_60.png   
...                                                 ...   
17220  https://cdn.sofifa.com/players/247/059/21_60.png   
17343  https://cdn.sofifa.com/players/251/734/21_60.png   
17659  https://cdn.sofifa.com/players/251/738/21_60.png   
17661  https://cdn.sofifa.com/players/251/741/21_60.png   
18205  https://cdn.sofifa.com/players/251/737/21_60.png   

                                               playerUrl  \
289    http://sofifa.com/player/230347/welington-kaue...   
292    http://sofifa.com/player/230225/juan-everton-m...   
369    http://sofifa.com/player/245299/jaime-nicolas-...   
374    http://sofifa.com/player/245294/jorge-ezequiel...   
375    http://sofifa.com/player/245308/mauro-evidio-n...   
...                                                  ...   
17220  http://sofifa.com/player/247059/ali-al-hidhani...   
17343  http://sofifa.com/player/251734/anudaan-sanghu...   
17659  http://sofifa.com/player/251738/madanapal-chad...   
17661  http://sofifa.com/player/251741/eelamynthan-su...   
18205  http://sofifa.com/player/251737/jagatjiva-mang...   

                Nationality  Age  OVA  POT     Club Contract_Status  \
289                  Brazil   20   81   81  No Club            Free   
292                  Brazil   24   81   81  No Club            Free   
369                 Uruguay   36   80   80  No Club            Free   
374                 Uruguay   32   80   80  No Club            Free   
375                 Uruguay   32   80   80  No Club            Free   
...                     ...  ...  ...  ...      ...             ...   
17220  United Arab Emirates   22   56   64  No Club            Free   
17343                 India   20   56   56  No Club            Free   
17659                 India   27   55   55  No Club            Free   
17661                 India   28   55   55  No Club            Free   
18205                 India   26   53   54  No Club            Free   

         Positions  Height_cm  Weight_kg Preferred_Foot  BOV Best_Position  \
289         LB, LM      178.0         69           Left   81            LB   
292        CB, CDM      181.0         82          Right   81            CB   
369        CB, CDM      181.0         82          Right   80            CB   
374             GK      190.0         85          Right   80            GK   
375         LB, LM      178.0         69           Left   80            LB   
...            ...        ...        ...            ...  ...           ...   
17220           RB      172.0         67          Right   56            RB   
17343       RM, CM      168.0         62          Right   56            RM   
17659           ST      178.0         77          Right   57            RM   
17661  RM, CAM, RB      175.0         73           Left   55            RM   
18205           LM      171.0         65           Left   53     

### Cleaning The Loan_Date_End Column

Before cleaning the column, lets analyse it carefully

In [43]:
#Getting the uniqu values in the column
data['Loan_Date_End'].unique()

array([nan, 'Jun 30, 2021', 'Dec 31, 2020', 'Jan 30, 2021',
       'Jun 30, 2022', 'May 31, 2021', 'Jul 5, 2021', 'Dec 31, 2021',
       'Jul 1, 2021', 'Jan 1, 2021', 'Aug 31, 2021', 'Jan 31, 2021',
       'Dec 30, 2021', 'Jun 23, 2021', 'Jan 3, 2021', 'Nov 27, 2021',
       'Jan 17, 2021', 'Jun 30, 2023', 'Jul 31, 2021', 'Nov 22, 2020',
       'May 31, 2022', 'Dec 30, 2020', 'Jan 4, 2021', 'Nov 30, 2020',
       'Aug 1, 2021'], dtype=object)

It is observed that the column contains NAN (Missing values) values

Let see how many NAN values is in the column

In [44]:
#Getting the number of Nan value in the Load_Date_End column
data['Loan_Date_End'].isnull().sum()

17966

There are 17966 missing value in the Load_date_End, since the amount of missing is far greater than the Available values, It is ideal to drop the entire column, Not all players are on loan, The Contract_Status column already justified that

In [45]:
# Removing the Loan_Date_End from the data
data = data.drop('Loan_Date_End', axis=1)

In [47]:
#Checking if our change is applied
data.head(4)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   

                                           playerUrl Nationality  Age  OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   91   

   POT             Club Contract_Status   Positions  Height_cm  Weight_kg  \
0   93     FC Barcelona        Contract  RW, ST, CF      170.0         72   
1   92         Juventus        Contract      ST, LW      187.0         83   
2   93  Atlético Madrid        Contract          GK      188.0         87   
3   91  Manchester City        Contract     CAM, CM      181.0         70   

  Preferred_Foot  BOV Best_Position        Joined    Value   Wage  \
0           Left   93            RW   Jul 1, 2004  €103.5M  €560K   
1          Right   92            ST  Jul 10, 2018     €63M  €220K   
2          Right   91            GK  Jul 16, 2014    €120M  €125K   
3          Right   91           CAM  Aug 30, 2015    €129M  €370K   

  Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   
2        €159.4M         95        13         11                15   
3          €161M        407        94         82                55   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   
2        140          34             19           11      65         11   
3        408          76             66           88      94         84   

   Composure  Defending  Marking  Standing_Tackle  Sliding_Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   
2         68         57       27               12              18   
3         91        186       68               65              53   

   Goalkeeping  GK_Diving  GK_Handling  GK_Kicking  GK_Positioning

Checking if the Preferred_Foot column is not messy

The column is all good

Since we have a very high number of missing values in the 'Loan_Date_End' column, It is best we drop the entire column

In [ ]:
data

#### Moving on to clean other columns

In [ ]:
data['Value']

we have unwanted variables, also we need to rename the column by assigning the currency type to it

In [ ]:
data.rename(columns = {'Value': 'Value_in_Euro'}, inplace = True)

In [ ]:
data.head(2)

We can now remove the euro sign from the entire values in the column

In [ ]:
data['Value_in_Euro'] = data['Value_in_Euro'].str.replace('€', '')

In [ ]:
data.head(3)

In [ ]:
data['Value_in_Euro'].unique()

After removing the units and renaming the columns, we can now convert the column into an int

In [ ]:
data['Value_in_Euro'].str[-1:] == 'K'

In [ ]:

#y = data['Value_in_Euro'].unique()
#strings = []
#numbers = []
if data['Value_in_Euro'].str[-1:] == 'K':  # Check if the last digit is K
    data['Value_in_Euro'].str[:-1] + "000" # Append a Thousand after removing the last char
#elif data['Value_in_Euro'].str[-1:] == 'M':  # Check if the last digit is M
#    data['Value_in_Euro'].str[:-1] + "000000"  # Append a Million after removing the last char
#else:  # just in case data doesnt have an M or K
#    data['Value_in_Euro'].astype(float)
        
        

In [ ]:
data[data['Value_in_Euro'].str[-1:] == 'K']

In [ ]:
data['Value_in_Euro']

In [ ]:
def convert_value(y):
    y = data['Value_in_Euro'].unique()
    strings = []
    numbers = []
    for number in y:
        if number[-1:] == 'K':  # Check if the last digit is K
            strings.append(number[:-1] + " Thousand")  # Append a Thousand after removing the last char
            numbers.append(float(number[:-1]) * 1000)  # Remove the last digit with [:-1], and convert to int and multiply by 1000
        elif number[-1:] == 'M':  # Check if the last digit is M
            strings.append(number[:-1] + " Million")  # Append a Million after removing the last char
            numbers.append(float(number[:-1]) * 1000000)  # Remove the last digit with [:-1], and convert to int and multiply by 1000000
        else:  # just in case data doesnt have an M or K
            strings.append(number)
            numbers.append(int(number))